crates/block-producer/src/mempool/mod.rs

```diff
@@ -0,0 +1,269 @@
+#![allow(unused)]
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    fmt::Display,
+    ops::Sub,
+    sync::Arc,
+};
+
+use batch_graph::BatchGraph;
+use miden_objects::{
+    accounts::AccountId,
+    transaction::{ProvenTransaction, TransactionId},
+    Digest,
+};
+use miden_tx::utils::collections::KvMap;
+use transaction_graph::TransactionGraph;
+
+use crate::store::TransactionInputs;
+
+mod batch_graph;
+mod transaction_graph;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchJobId(u64);
+
+impl Display for BatchJobId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BatchJobId {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockNumber(u32);
+
+impl Display for BlockNumber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BlockNumber {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+
+    pub fn checked_sub(&self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(Self)
+    }
+}
+
+pub struct Mempool {
+    /// The latest inflight state of each account.
+    ///
+    /// Accounts without inflight transactions are not stored.
+    account_state: BTreeMap<AccountId, (Digest, TransactionId)>,
+
+    /// The number of the next block we hand out.
+    next_inflight_block: BlockNumber,
+
+    /// Inflight transactions.
+    transactions: TransactionGraph,
+
+    /// Inflight batches.
+    batches: BatchGraph,
+
+    /// The next batches ID.
+    next_batch_id: BatchJobId,
+
+    /// Blocks which are inflight or completed but not yet considered stale.
+    block_pool: BTreeMap<BlockNumber, Vec<BatchJobId>>,
```

Comment:
This is for keeping track of recent blocks, right? If so, I'm wondering if we actually need this. Specifically, there are 3 things from recent blocks that we need to keep track of:

- Account states - basically `(AccountId, Digest)` tuples.
- Newly created notes - `NoteId`s, maybe with inclusion paths.
- Newly created nullifiers - these would be just `Nullifier` structs stored in a set or something like that.

As you work though adding note/nullifier tracking to the mempool, it might make sense to store them together and get rid of this `block_pool` field.

Reply:
These track all inflight blocks as well as recently applied blocks. The only reason we need the recent blocks is to efficiently trim the current state once we consider them beyond the "stale" threshold of incoming transactions. Otherwise the in-memory state just grows indefinitely.

This would also be used if we have multiple inflight blocks and one fails. So I'm not sure if we do ever want to remove this? Unless we enforce that there is at most exactly one block inflight.

Reply:
> This would also be used if we have multiple inflight blocks and one fails. So I'm not sure if we do ever want to remove this? Unless we enforce that there is at most exactly one block inflight.

As mentioned in another comment, I would not allow multiple in-flight blocks for now.

> The only reason we need the recent blocks is to efficiently trim the current state once we consider them beyond the "stale" threshold of incoming transactions.

I think we should be able to do the same without tracking the actual blocks. For example, as soon as a block is committed, we'd remove all relevant data from transaction and batch graph (this shouldn't require examining the entire state). Then, we'd update the "in-flight" data structure to add block number of the recently committed block to relevant pieces of data (account updates, nullifiers, notes). At this time, we'd also remove the stale data associated with some old block.

The above really depends on how we define the structure for the in-flight data though.

Reply:
Resolving this thread in favor of this related one [here](https://github.com/0xPolygonMiden/miden-node/pull/502#discussion_r1783830150) to keep discussing the state tracking options.